### PR TITLE
Improve tests, replace psr7 implementation

### PR DESCRIPTION
**Pull request by acasademont**
Trying to add tests for #144 proved to be quite complicated if we don't have a full Symfony kernel for the tests. Therefore, I've refactored all the current tests to stop using mocks and use a real Symfony kernel/controllers...

As a side effect, the current PSR-7 library did not support the `UploadFileInterface` and replace it with Guzzle (which is better maintained by the way).

This is just the first step in all the planned refactors :)

### CODE REVIEW

File: Bootstraps/Symfony.php [hunk 1]
```diff
@@ -88,7 +89,7 @@ public function getApplication()
             $app->booted = true;
         }, $app);
 
-        if ($this->debug) {
+        if ($this->debug && ProcessSlave::$slave) {
```
Comment (acasademont) [inline Bootstraps/Symfony.php]: The tests don't use a PHP-PM slave and the `register_file` function failed miserably. Not sure if we should fix the function instead to check for the `$slave` property not being null

File: Bootstraps/Symfony.php [hunk 2]
```diff
@@ -88,7 +89,7 @@ public function getApplication()
             $app->booted = true;
         }, $app);
 
-        if ($this->debug) {
+        if ($this->debug && ProcessSlave::$slave !== null) {
```
Comment (acasademont) [inline Bootstraps/Symfony.php]: Weird side effect, when running the tests we have no Slave process and the `register_file` function fails miserably, not sure if we should fix the function instead...

File: composer.json [hunk 3]
```diff
@@ -3,12 +3,14 @@
     "license": "MIT",
     "require": {
         "php-pm/php-pm": "^1.0.5",
-        "symfony/http-foundation": "^2.6|^3.0|^4",
-        "symfony/http-kernel": "^2.6|^3.0|^4",
-        "ringcentral/psr7": "^1.2"
+        "symfony/http-foundation": "^3.4|^4",
```
Comment (acasademont) [inline composer.json]: Dropped 2.X already

**Comment by andig** [comment]
Wow, this is huge! Will need some time to check, sorry.

**Comment by acasademont** [comment]
No worries, happy to help! The test assertions haven't changed, seems big but I don't think is thaaat big :p

File: tests/SymfonyMocks/Kernel.php [hunk 4]
```diff
@@ -1,87 +0,0 @@
-<?php
```
Comment (acasademont) [inline tests/SymfonyMocks/Kernel.php]: @andig basically I copied all the code in the previous Kernel mock and put it inside the newly created controllers, that allows us to keep the assertions the same

**Comment by andig** [comment]
As for symfony bundle: is it a good idea to require them in the composer json? That way we will only ever be able to install a single framework instead of testing them separately. Wondering if it would make more sense to move the symfony framework to travis.yml to prepare a separate environment per framework.

Apart from that LGTM.

File: composer.json [hunk 5]
```diff
@@ -3,12 +3,14 @@
...
+        "symfony/http-kernel": "^3.4|^4",
+        "guzzlehttp/psr7": "^1.5"
     },
     "require-dev": {
-        "phpunit/phpunit": "^5.7"
+        "phpunit/phpunit": "^5.7",
+        "symfony/framework-bundle": "^3.4|^4",
+        "doctrine/annotations": "^1.6"
```
[the remark this answers is not in the document]
Comment (acasademont) [reply composer.json]: Nopes, they aren't implicit, Symfony relies on them for routing annotation but you need to import the library yourself

**Comment by acasademont** [comment]
> As for symfony bundle: is it a good idea to require them in the composer json? That way we will only ever be able to install a single framework instead of testing them separately. Wondering if it would make more sense to move the symfony framework to travis.yml to prepare a separate environment per framework.

It's a good question! The three frameworks can live together without a problem, we're already doing that for phpstan in travis (we install Laravel, Symfony and Drupal). I'm not a Laravel or Drupal expert but I don't see why testing them should interfere with Symfony. 

Also, I'm not a fan of moving too much logic into Travis, imo tests should be runnable with a simple "phpunit" command.

**Comment by andig** [comment]
Then I would propose to merge as-is and then proceed with #144

**Comment by acasademont** [comment]
@andig sounds good to me! I will submit another PR before #144 that adds new tests for all the hacks we have in place and then check if the tests still pass after the refactor.

**Pull request merged**